### PR TITLE
sshx: init at unstable-2023-11-04

### PR DESCRIPTION
**Pull request by pinpox**
## Description of changes

Init sshx, A secure web-based, collaborative terminal

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/misc/sshx/default.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    darwin.apple_sdk.frameworks.Security
+  ];
+
+  env = {
+    ZSTD_SYS_USE_PKG_CONFIG = true;
+  };
+
+  meta = with lib; {
+    description = "Fast, collaborative live terminal sharing over the web";
+    homepage = "git@github.com:ekzhang/sshx.git";
```
Comment (Mic92) [inline pkgs/tools/misc/sshx/default.nix:41]: Copy and paste error?

```suggestion
    homepage = "https://github.com/ekzhang/sshx";
```

**Comment by MikaelFangel** [comment]
This package should be added under the `pkgs/by-name/so/some-example/package.nix` if possible as it's the new preferred way to add packages (:

**Comment by pinpox** [comment]
> This package should be added under the `pkgs/by-name/so/some-example/package.nix` if possible as it's the new preferred way to add packages (:

Where can I find details on that? Searching [the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/) for `by-name` yields no results. Is the format the same? Do I have to still add them to `all-packages.nix` aswell?

**Comment by MikaelFangel** [comment]
You can read about it here:
[https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md) and the readme under pkgs also uses a by-name example (:

**Comment by MikaelFangel** [comment]
And regarding the pr in general you should rebase it so it only contains one commit:

`sshx: init at unstable-x.x.x`

**Comment by pinpox** [comment]
> And regarding the pr in general you should rebase it so it only contains one commit:

Squashed now. Thanks for the info!

**Comment by NyCodeGHG** [comment]
Maybe it would be a good idea to separate the server and client binaries into their own outputs, so users don't have to download the server if they only want to use the client and vice-versa?

**Comment by MikaelFangel** [comment]
@NyCodeGHG - As I see it, sshx is the server program and clients are just browsers which join via urls. Or maybe I've missed something in the repo that makes this kind of configuration possible?

Notice how it says 

> Run this command in your favorite terminal.
>`sshx`
>This kicks off a live, encrypted session. Open the link in your web browser to join.

in step 2: [https://sshx.io/](https://sshx.io/)

**Comment by NyCodeGHG** [comment]
sshx is the client program which talks to a hosted server (https://sshx.io by default), the server binary is sshx-server which is also built in this expression but not separated

**Comment by MikaelFangel** [comment]
Oh, totally missed that, thanks. Then I agree in your suggestion 👍🏻

**Comment by pinpox** [comment]
> Maybe it would be a good idea to separate the server and client binaries into their own outputs, so users don't have to download the server if they only want to use the client and vice-versa?

Sorry, but I don't know how to do that. Can anyone point me to a golang example with multiple outputs?

**Comment by NyCodeGHG** [comment]
```nix
outputs = [ "out" "server" ];

postInstall = ''
  moveToOutput 'bin/sshx' "$out"
  moveToOutput 'bin/sshx-server' "$server"
'';
```
adding this should be enough

But i think to actually run the server building the frontend would also be required, but i think we can postpone that [until self hosting the server is actually supported](https://github.com/ekzhang/sshx/pull/45#issuecomment-1806824666).

**Comment by pinpox** [comment]
> But i think to actually run the server building the frontend would also be required, but i think we can postpone that [until self hosting the server is actually supported](https://github.com/ekzhang/sshx/pull/45#issuecomment-1806824666).

Outputs are separate now. Kind of a bummer though, that the server is not self-hostable yet, I didn't know that.

**Comment by NyCodeGHG** [comment]
I already tried selfhosting, but gave up after weird issues with gRPC and nginx :/

**Comment by pinpox** [comment]
> I already tried selfhosting, but gave up after weird issues with gRPC and nginx :/

Ok, can we merge this anyway? I find it still useful with the hosted server as is

**Comment by Kranzes** [comment]
> I already tried selfhosting, but gave up after weird issues with gRPC and nginx :/

check out https://github.com/NixOS/nixpkgs/pull/271881 and https://github.com/NixOS/nixpkgs/pull/271948 :^)

**Pull request merged**